### PR TITLE
MCO-707: Strip registry transport from os image URL for comparison

### PR DESCRIPTION
**Pull request by ori-amizur**
One of the conditions to reboot after firstboot is different os image
URLs.  In order to correctly compare the URLs, the booted OS image URL
has to be stripped from container image reference and transport.
This commit utilizes a recent change in rpm-ostree client to do just
that.

**- What I did**
Strip registry transport from os image URLs before checking if reboot is needed.
**- How to verify it**
1. Regression
2. Verify if URLs are considered equal when booted os URL is with a registry transport, and the required is without.
**- Description for the changelog**

Strip registry transport from OS Image URLs for comparison

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -2075,6 +2075,10 @@ func (dn *Daemon) validateOnDiskState(currentConfig *mcfgv1.MachineConfig) error
 	return nil
 }
 
+func canonizeImageURL(url string) string {
```
Comment (cgwalters) [inline pkg/daemon/daemon.go:2078]: This could use a doc comment (and also "canonicalize" is more canonical here) - [canonization is a different thing](https://en.wikipedia.org/wiki/Canonization) and I don't think this function quite qualifies for sainthood yet :smile: .  But even "canonicalize" would make someone wonder "canonicalize *how*?" I think.  

Further this is going to be a bit fragile because in theory we could (and in fact [have a TODO to change it](https://github.com/ostreedev/ostree-rs-ext/blob/332ad0dd799edbe023472a47c427ce81954b0fc5/lib/src/container/mod.rs#L240)).   So how about combining these two:  
```suggestion
// stripRegistryTransport removes the registry transport prefix, see https://docs.rs/ostree-ext/latest/ostree_ext/container/enum.Transport.html
func stripRegistryTransport(url string) string {
        return strings.Relpace(strings.Replace(url, "docker://", "", 1), "registry:", "", 1))
```
or so?

Could also use a unit test.

Comment (sinnykumari) [reply pkg/daemon/daemon.go:2078]: Do we know that it will be always "docker://" or we should handle rest of the transport replacement as well that are supported by [OSTree](https://docs.rs/ostree-ext/latest/ostree_ext/container/enum.Transport.html) ?

**Review by sinnykumari — COMMENTED** [review]
One question, other than that I agree with changes already suggested by Colin.

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (cgwalters) [reply pkg/daemon/daemon.go:2078]: I think in OpenShift today we effectively hardcode `docker://` - i.e. we only support fetching from a registry in `osImageURL`.

So...yes.  Note https://github.com/openshift/os/pull/657 would actually switch things so that the initial state is from an `oci-archive`, but that's OK because it will today always be older anyways.

(A whole other problem here is that today we deploy as OCI, but when we later push the container image it gets converted to d2s2, which changes the digest...)

Comment (sinnykumari) [reply pkg/daemon/daemon.go]: yeah, makes sense.

File: pkg/daemon/update_test.go [hunk 2]
```diff
@@ -225,6 +225,26 @@ func TestMachineConfigDiff(t *testing.T) {
...
+	assert.Nil(t, err)
+	assert.True(t, diff.isEmpty())
+
+	oldConfig.Spec.OSImageURL = "registryy:" + osImageURL
+	newConfig.Spec.OSImageURL = osImageURL
+	diff, err = newMachineConfigDiff(oldConfig, newConfig)
+	assert.Nil(t, err)
+	assert.False(t, diff.isEmpty())
+	assert.True(t, diff.osUpdate)
+
```
Comment (sinnykumari) [inline pkg/daemon/update_test.go:247]: Can we also add a test case where oldConfig has empty OSImageURL and newConfig has a valid OSImageURL. something like oldConfig.Spec.OSImageURL = " "  and newConfig.Spec.OSImageURL = ""quay.io/example/foo@sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c" .

This usually is the case today during initial node bootstrap for a cluster.

**Comment by sinnykumari** [comment]
Also, will be good to reword commit message based on latest changes that we have.

File: pkg/daemon/update_test.go (hunk 2 above)
Comment (ori-amizur) [reply pkg/daemon/update_test.go:247]: Doesn't this already exist on line 222?

Comment (sinnykumari) [reply pkg/daemon/update_test.go:247]: ah yes, should have looked at existing test cases.

**Comment by sinnykumari** [comment]
Thanks Ori for working on this.
/lgtm
This should get merged once 4.14 branching has happened and we no longer require valid-bz label.

**Comment by sergiordlr** [comment]
Verifying using IPI on GCP

1. Create a custom osImage (we used the internal registry)
`image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/tc-54056-new-os-image@sha256:5ccada2081859a95414cd3f6a96a5f871e9ef8b5463ad28e4e8128bfe503faa3`


2. Deploy a mc to use this custom osImage

```
kind: MachineConfig
apiVersion: machineconfiguration.openshift.io/v1
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: "test-new-os-image"
spec:
  osImageURL: "image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/tc-54056-new-os-image@sha256:5ccada2081859a95414cd3f6a96a5f871e9ef8b5463ad28e4e8128bfe503faa3"
```

The new osImage is applied and the nodes rebooted.

3. Edit the test-new-os-image MC create in step 2 to prepend "docker://" to the osIMage. Like this:

```
spec:
  osImageURL: "docker://image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/tc-54056-new-os-image@sha256:5ccada2081859a95414cd3f6a96a5f871e9ef8b5463ad28e4e8128bfe503faa3"

```

A new MC is rendered and, when it is applied to the nodes, it is considered to be already applied and the nodes are not rebooted nor drained, as expected.

Nevertheless, we need to be aware that  "docker://..." is not right format for osImageURL. So we are accepting a wrong format in  osImageURL if the image is the same as the one already present in the cluster. Imho, It's not a big deal, but we need to be aware of it. A customer could wonder why "docker://" is accepted in the MCs' osImageURL only sometimes.

Regarding the agent-intaller, we can't verify this PR using an agent-installer until https://github.com/openshift/os/pull/657 is merged. Is it right?

**Comment by rioliu-rh** [comment]
hold this PR according to Sergio's comment, another issue about negative case, if we create a mc to update os image with a new one w/o transport prefix `docker://`, the pool will be degraded with message like `docker://... is not the right format`

Sergio can provide more details about it.

**Comment by cgwalters** [comment]
> Nevertheless, we need to be aware that "docker://..." is not the right format for osImageURL. So we are accepting a wrong format in osImageURL if the image is the same as the one already present in the cluster. Imho, It's not a big deal, but we need to be aware of it. A customer could wonder why "docker://" is accepted in the MCs' osImageURL only sometimes.

Hmm yes.  I think we should actually continue to reject `docker://` being present in `osImageURL`.  I don't see what we gain by allowing it there.

So this PR started with:
> sometimes the URL starts with a prefix (i.e docker:// or registry:), and sometimes without

Let's try to identify more precisely the code flows involved here.

I'm actually a bit confused here because I thought we were parsing the output from `rpm-ostree status --json`, which has e.g.:
`"container-image-reference" : "ostree-unverified-registry:quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:cd29255bd4e6634404ef8e0f4a7f19b4ae5a57a3afc7d67cbe00934ba998d6ac"`

And then we sort of hackily scrape that off...

Ohh, wait hmm...I think I may understand now!  Yes, the bug is in that code in the MCO that's incorrectly parsing ostree-container image references.  When we use `ostree container image deploy`, it ends up as e.g.:

```
$ rpm-ostree status --json | jq -r '.deployments[0]."container-image-reference"'
ostree-unverified-image:docker://quay.io/fedora/fedora-coreos:stable
```

So the core bug is here https://github.com/openshift/machine-config-operator/blob/5b821a279c88fee1cc1886a6cf1ec774891a2258/pkg/daemon/rpm-ostree.go#L155

Now an immediate practical problem is that the code for parsing these strings today lives in Rust, not Go.  (And even for using Rust, we don't really want to vendor all of ostree-rs-ext, which is its own distinct bug).

Here's what I'd propose:

- Add a minimal Go reimplementation of the logic in https://github.com/ostreedev/ostree-rs-ext/blob/8b04520d8e0d3ffb0c6ba72098d4114083ee7124/lib/src/container/mod.rs#L183 into https://github.com/coreos/rpmostree-client-go/
- Bump that dependency and use it here instead of just stripping from the first `:`
-

**Comment by cgwalters** [comment]
I can also look at "canonicalizing" `ostree-unverified-image:docker://` back to `ostree-unverified-registry:` on the ostree side, which would also implicitly fix things.

**Comment by cgwalters** [comment]
/lgtm cancel
per above

**Comment by cgwalters** [comment]
https://github.com/ostreedev/ostree-rs-ext/pull/533

**Comment by ori-amizur** [comment]
> Here's what I'd propose:
> 
> * Add a minimal Go reimplementation of the logic in https://github.com/ostreedev/ostree-rs-ext/blob/8b04520d8e0d3ffb0c6ba72098d4114083ee7124/lib/src/container/mod.rs#L183 into https://github.com/coreos/rpmostree-client-go/
> * Bump that dependency and use it here instead of just stripping from the first `:`

How would this solve the problem of comparing URLs with the registry transports? Would the change in the rpmsostree-client handle the stripping of the transports too?

**Comment by cgwalters** [comment]
PR in https://github.com/coreos/rpmostree-client-go/pull/21

**Comment by cgwalters** [comment]
So next we need https://github.com/openshift/machine-config-operator/pull/3916
and then we can update this PR to execute on the bottom half of https://github.com/openshift/machine-config-operator/pull/3857#issuecomment-1712021526

**Comment by sinnykumari** [comment]
https://github.com/openshift/machine-config-operator/pull/3916 has been merged, we should be good to utilize bootedDeployment.RequireContainerImage() to get bootedOSImageURL without worrying about transport prefix.

**Comment by sinnykumari** [comment]
LGTM
/approve
@sergiordlr This should be ready for qe testing now
@cgwalters PTAL

**Comment by sergiordlr** [comment]
Hello! @sinnykumari this PR depends on https://github.com/openshift/os/pull/657

We need this PR to be merged before we can execute the verification here.

**Comment by sinnykumari** [comment]
Thanks Sergio for testing it. Agree it is difficult to test complete functionality without https://github.com/openshift/os/pull/657 . With current testing, it should be sufficient for the PR to get merged as it doesn't break any existing functionalities. 
/lgtm
